src/utils/fetch-source.js

```diff
@@ -89,15 +99,68 @@ export async function fetchSource(
       }
       break;
     }
+
     case 'CORS_ENABLED_ARCHIVE': {
-      // Not handled yet.
-      errors.push({ type: 'NO_KNOWN_CORS_URL' });
+      // Download the archive, and then get the file from the archive.
+      // This is the path for source code from crates.io: We download the entire
+      // crate, which is a .tar.gz file, and then we store the uncompressed bytes
+      // in the archive cache.
+      // We only download each archive once, and reuse the stored bytes when we
+      // want to show other files from the same archive.
+
+      const { archiveUrl: url, pathInArchive } = downloadRecipe;
+      let promise: Promise<Uint8Array> | void = archiveCache.get(url);
+
+      if (promise === undefined) {
+        // Create a promise to load the archive, but don't await it yet.
+        promise = (async () => {
+          const response = await callbacks.fetchUrlResponse(url);
+          const bytes = new Uint8Array(await response.arrayBuffer());
+          return isGzip(bytes) ? decompress(bytes) : bytes;
+        })();
+        archiveCache.set(url, promise);
+      } else {
+        // TODO: callbacks.onWaitingForUrlWhichIsAlreadyBeingFetched(url)
+      }
+
+      try {
+        // Now we await the promise. If we only just started loading the archive,
+        // then this is where we wait for the network load to complete.
+        // If the archive has already finished loading, then this is just a
+        // microtask round-trip.
+        const bytes = await promise;
+
+        // Find the file inside of the archive.
+        const stream = new UntarFileStream(bytes.buffer);
+        const textDecoder = new TextDecoder();
```

Comment:
For this `TextDecoder` instance, would that be a good idea to indicate `{ fatal: true }` in the constructor so that it throws when it can't decode the text (as opposed to producing garbled characters)?

(mdn documentation for the constructor: https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/TextDecoder)

Reply:
Hmm, I think for this one I would actually use non-fatal parsing. This is where we obtain the source code. I prefer source code with a few garbled characters over no source code at all.
It's not like we're trying to execute that source code and need it to be 100% correct.